### PR TITLE
104 full training mode

### LINKED ISSUE #104: "full training" mode
### New Feature Summary

Current trainer code performs k-fold evaluation during training, and thus requires split size as a part of trainer configuration 

https://github.com/clamsproject/app-swt-detection/blob/5925f029c6f13446a78e7144ae30f146354186a2/modeling/train.py#L164

I'd like to have a "full training" mode in the trainer code that _when `num_split==1`, is does train a model given the rest of configuration, without k-fold evaluation, and save the checkpoint file, without printin/saving P/R/F scores. This will be helpful to train a model that uses the maximum amount of training data, once the hyperparameters are set after k-fold based experiments. 

### Related

_No response_

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by jyoune**
Added "full training mode" which involves an if condition in the `k_fold_train` method. If `config['num_splits']` is set to 1, then the model is trained with all the documents at once, instead of doing k folds. There is also no evaluation, as the validation set is empty (everything in train). fixes #104 full training mode

### CODE REVIEW

File: modeling/train.py [hunk 1]
```diff
@@ -161,42 +161,64 @@ def k_fold_train(indir, outdir, config_file, configs, train_id=time.strftime("%Y
...
-        if not train.has_data() or not valid.has_data():
-            logger.info(f"Skipping fold {i} due to lack of data")
-            continue
-        train_loader = DataLoader(train, batch_size=40, shuffle=True)
-        valid_loader = DataLoader(valid, batch_size=len(valid), shuffle=True)
+        train_loader = DataLoader(train, batch_size=len(guids), shuffle=True)
         loss = nn.CrossEntropyLoss(reduction="none")
         device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
-        logger.info(f'Split {i}: training on {len(train_guids)} videos, validating on {validation_guids}')
         export_csv_file = f"{outdir}/{train_id}.kfold_{i:03d}.csv"
```
Comment (keighrim) [inline modeling/train.py:214]: clean up unused variables

File: modeling/train.py [hunk 2]
```diff
@@ -161,42 +161,64 @@ def k_fold_train(indir, outdir, config_file, configs, train_id=time.strftime("%Y
...
         export_model_file = f"{outdir}/{train_id}.kfold_{i:03d}.pt"
```
Comment (keighrim) [inline modeling/train.py:215]: Suggesting changing the output checkpoint file name, namely get rid of .kfold_xxx` suffix, since it's no longer "kfold" when in full train mode. (then, `i` variable won't be necessary)

File: modeling/train.py [hunk 3]
```diff
@@ -161,42 +161,64 @@ def k_fold_train(indir, outdir, config_file, configs, train_id=time.strftime("%Y
...
+        p_config = Path(f'{outdir}/{train_id}.kfold_config.yml')
+        p_results = Path(f'{outdir}/{train_id}.kfold_results.txt')
+        p_results.parent.mkdir(parents=True, exist_ok=True)
+        export_kfold_config(config_file, configs, p_config)
+        export_kfold_results(val_set_spec, p_scores, r_scores, f_scores, p_results, **configs)
+    # if num_splits == 1. validation is empty.
+    else:
+        i = 1
+        train_guids = set(guids)
+        validation_guids = set([])
```
Comment (keighrim) [inline modeling/train.py:206]: clean up unused variables

File: modeling/train.py [hunk 4]
```diff
@@ -161,42 +161,64 @@ def k_fold_train(indir, outdir, config_file, configs, train_id=time.strftime("%Y
...
+        # prepare_datasets seems to work fine with empty validation set
         train, valid, labelset_size = prepare_datasets(indir, train_guids, validation_guids, configs)
-        # `train` and `valid` vectors DO contain positional encoding after `split_dataset`
-        if not train.has_data() or not valid.has_data():
-            logger.info(f"Skipping fold {i} due to lack of data")
-            continue
-        train_loader = DataLoader(train, batch_size=40, shuffle=True)
-        valid_loader = DataLoader(valid, batch_size=len(valid), shuffle=True)
+        train_loader = DataLoader(train, batch_size=len(guids), shuffle=True)
         loss = nn.CrossEntropyLoss(reduction="none")
```
Comment (keighrim) [inline modeling/train.py:212]: `loss = `
and 
`device = ` 
lines can be factored out of the if conditional..?

**Review by keighrim — CHANGES_REQUESTED** [review]
Functionality-wise, the implementation seems to be spot-on. Just suggesting some changes for optimizing code.

**Review by keighrim — APPROVED** [review]
